### PR TITLE
feat(graphql): expose specs via graphql

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes cypress-io.atlassian.net/browse/UNIFY-301 
- Closes cypress-io.atlassian.net/browse/UNIFY-302

Expose specs via GraphQL using Nexus Relay Connection plugin. Docs: https://nexusjs.org/docs/plugins/connection

I am very glad we are using the Nexus object syntax, which allows us to easily handle GraphQL Connections using the Relay plugin. It's great.

Example:

image

### CODE REVIEW

File: packages/server/lib/gui/events.ts [hunk 1]
```diff
@@ -536,6 +537,9 @@ module.exports = {
         getProjectRootsFromCache () {
           return getProjectRoots()
         },
+        findSpecs (payload: FindSpecs) {
```
Comment (lmiller1990) [inline packages/server/lib/gui/events.ts:540]: Injecting the old, untyped code - I like this pattern.

File: packages/data-context/src/data/loaders.ts [hunk 2]
```diff
@@ -28,6 +29,14 @@ export class DataLoaders {
     return this.configLoader.load(projectRoot)
   }
 
+  gitInfo (path: string) {
+    return this.gitLoader.load(path)
+  }
+
+  private gitLoader = this.loader<string, GitInfo>((absolutePaths) => {
+    return getGitInfo(absolutePaths)
```
Comment (tgriesser) [inline packages/data-context/src/data/loaders.ts:37]: We'll need to account for invalidation here, but we can do that separately

**Review by tgriesser — APPROVED** [review]
Nice! Added a few comments but I think we're good to merge and we can add those changes separately

File: packages/data-context/src/data/loaders.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/data-context/src/data/loaders.ts:37]: Yeah, I am guessing this is part of the general topic of "how we watch files in Cypress".

File: packages/graphql/src/schemaTypes/objectTypes/gql-Project.ts [hunk 3]
```diff
@@ -44,5 +45,17 @@ export const Project = objectType({
         return ctx.project.isFirstTimeAccessing(source.projectRoot, 'e2e')
       },
     })
+
+    t.connectionField('specs', {
+      description: 'Specs for a project conforming to Relay Connection specification',
+      type: 'Spec',
+      additionalArgs: {
+        specType: arg({ type: SpecTypeEnum }),
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Project.ts:53]: removed it anyway

File: packages/graphql/src/schemaTypes/objectTypes/gql-GitInfo.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { objectType } from 'nexus'
+
+export const GitInfo = objectType({
```
Comment (ImCesar) [inline packages/graphql/src/schemaTypes/objectTypes/gql-GitInfo.ts:3]: This is subjective but I like the name `Git` over `GitInfo`. I think the Info is implied by it giving us data and Git lines up with our other nodes like `Project` and `Spec`

**Review by ImCesar — COMMENTED** [review]
Just a naming suggestion, the rest looks good to me

File: packages/graphql/src/schemaTypes/objectTypes/gql-GitInfo.ts (hunk 4 above)
Comment (tgriesser) [reply packages/graphql/src/schemaTypes/objectTypes/gql-GitInfo.ts:3]: I think `GitInfo` makes sense for this, or even `SpecGitInfo`. I see what you're getting at @ImCesar , but `Project` / `Spec` are first class data models of the app, but `Git` makes it sound like it'd be a more generic interface to exposing data from the `git`:

```graphql
git {
   lastCommit
   untrackedFileCount
   isDirty
   currentBranch
}
```

**Pull request merged**